hack/bin/Sbom.hs

```diff
@@ -0,0 +1,376 @@
+  - string contexts like
+    ```nix
+    x = /* bash */ ''
+      cp ${pkgs.bla}/bin $out
+    '';
+    ```
+    would ignore dependencies on `pkgs.bla`
+- we can build the dependency graph independently (without knowing about the meta) but we somehow need
+  to obtain the meta itself
+- people don't always have a complete package set but more commonly are hand assembling things; we need
+  to give the possibility to build meta "databases" from package sets
+- we need to trace which dependencies are missing when querying the meta database against them
+- collecting the meta also poses some issue
+  - nixpkgs is not a tree, but a more general graph
+  - it also not a DAG but it has loops
+  - this means more specifically that we cannot without care recurse into it
+  - even if we only recurse very shallowly, we soon start running out of memory, this means we probably need
+    to do some on the fly filtering by "actual" dependencies
+  - this is similarly an issue, because it means that for every package we have to evaluate the entirety
+    of the package set instead of being able to keep and persist the database
+  - a more clean solution would probably be to at each time we recurse, a derivation that does the evaluation
+    and outputs a JSON that can later be read
+
+how this relates to bombon:
+- bombon uses a more coarse grained approach
+- this builds a metadata "database" i.e. is two pass
+- see the corresponding nix code in ./nix
+-}
+
+module Sbom where
+
+import Control.Arrow ((&&&))
+import Data.Aeson
+import Data.Aeson.Key qualified as KM
+import Data.Aeson.KeyMap qualified as KM
+import Data.Aeson.Types (typeMismatch)
+import Data.Bifunctor (first)
+import Data.Bitraversable (bitraverse)
+import Data.ByteString (ByteString)
+import Data.ByteString.Char8 qualified as C8
+import Data.ByteString.Lazy (LazyByteString)
+import Data.ByteString.Lazy qualified as BSL
+import Data.ByteString.Lazy.Char8 qualified as C8L
+import Data.Containers.ListUtils (nubOrd, nubOrdOn)
+import Data.Functor.Identity
+import Data.Map (Map)
+import Data.Map qualified as M
+import Data.Maybe
+import Data.Proxy
+import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Text.IO qualified as T
+import Data.Time.Clock.POSIX
+import Data.Traversable (for)
+import Data.Tree
+import Data.UUID qualified as UUID
+import Data.UUID.V4 qualified as V4
+import Debug.Trace
+import GHC.Generics hiding (Meta)
+import GHC.IsList (IsList (fromList, toList))
+import Numeric.Natural (Natural)
+import Options.Applicative (customExecParser, fullDesc, help, long, prefs, progDesc, showHelpOnEmpty, strOption, value)
+import Options.Applicative qualified as Opt
+import System.Directory
+import System.Process
+
+data License = MkLicense
+  { id :: Maybe Text,
+    name :: Maybe Text
+  }
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving anyclass (FromJSON, ToJSON)
+
+sadSbomMeta :: Text -> Text -> [Text] -> SBomMeta Identity
```

Comment:
```suggestion
emptySbomMeta :: Text -> Text -> [Text] -> SBomMeta Identity
```
or
```suggestion
notFoundSbomMeta :: Text -> Text -> [Text] -> SBomMeta Identity
```
?

Hm, could be that I don't get the name :thinking: However, if this is named "sad" because the search algorithm is sad because it didn't find anything, then I would say be can name this better :smiley:

Reply:
yeah, it's because it didn't find anything and the information it can provide is sad :(